### PR TITLE
Switch to simpler ssh-agent.

### PR DESCRIPTION
**Pull request by ellisonmarks**

### CODE REVIEW

File: lib/ssh/setup-ssh-agent [hunk 1]
```diff
@@ -0,0 +1,9 @@
+sshdir=$(getent passwd $USER | cut -d: -f6)/.ssh
+mkdir $sshdir
+chmod 700 $sshdir
+cat > $sshdir/key
+chmod 600 $sshdir/key
+eval "$(ssh-agent -s)"
```
Comment (bukzor) [inline lib/ssh/setup-ssh-agent:6]: sshdir is allowed to be group-other readable -- people are meant to be able to see your pubkeys. It's only writability that's a concern there.

To write the key without group-other readability even during race conditions:


```suggestion
mkdir -p "$sshdir"
umask 077
cat > "$sshdir/key"
eval "$(ssh-agent)"
```

([-s is default](https://www.man7.org/linux/man-pages/man1/ssh-agent.1.html); `mkdir -p` is all that's needed to make this script idempotent)

It obviously doesn't matter much here, but it's a good habit.

P3 or even P4

File: lib/ssh/setup-ssh-agent [hunk 2]
```diff
@@ -0,0 +1,9 @@
...
+gha-set-env SSH_AUTH_SOCK <<< $SSH_AUTH_SOCK
+gha-set-env SSH_AGENT_PID <<< $SSH_AGENT_PID
+ssh-add $sshdir/key
```
Comment (bukzor) [inline lib/ssh/setup-ssh-agent:9]: If you're going to give an argument to `ssh-add` (I honestly forgot you could do that) then you don't actually need to mess with sshdir at all, or even write the private key to disk, which actually seems preferable.


```
ssh-add -  # activate private key on stdin
```

Comment (bukzor) [inline lib/ssh/setup-ssh-agent:9]: The `gha` commands in this script means its scope is smaller than ssh. I think it belongs best under github-actions?

All together:

`lib/github-actions/ssh-agent`
```suggestion
#!/bin/bash
set -euxo pipefail
eval "$(ssh-agent)"
ssh-add -  # private key on stdin
gha-set-env SSH_AUTH_SOCK <<< "$SSH_AUTH_SOCK"
gha-set-env SSH_AGENT_PID <<< "$SSH_AGENT_PID"
```

`bin/gha-ssh-agent -> ../lib/github-actions/ssh-agent`

File: .github/actions/set-username-and-hostname/action.yml [hunk 3]
```diff
@@ -43,7 +43,7 @@ runs:
         useradd -D
 
         : add the requested username
-        sudo useradd -g "$olduser" -G "$(id -G | tr ' ' ,)" "$newuser"
+        sudo useradd -m -g "$olduser" -G "$(id -G | tr ' ' ,)" "$newuser"
```
[the remark this answers is not in the document]
Comment (bukzor) [reply .github/actions/set-username-and-hostname/action.yml:46]: I think things will work with a minimum number of edge cases if instead we symlink the two user's homedirs together. I've taken pains (elsewhere) to ensure most everything in `/home/runner` is read/writable by `$newuser`.

```bash
newhome="$(getent passwd "$newuser" | cut -d: -f6)"
sudo ln -sf "$HOME" "$newhome" 
```

**Review by bukzor — CHANGES_REQUESTED** [review]
Thanks!

It looks like this turned out as simple as I had hoped. I'm glad.
Sorry about all the comments =X
The later comments invalidate the earlier ones, a bit, so be sure to skim through before you change anything.

File: .github/actions/set-username-and-hostname/action.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (ellisonmarks) [reply .github/actions/set-username-and-hostname/action.yml]: made it use long options, but also removed create-home and set the symlink.

File: lib/ssh/setup-ssh-agent [hunk 4]
```diff
@@ -0,0 +1,9 @@
+sshdir=$(getent passwd $USER | cut -d: -f6)/.ssh
+mkdir $sshdir
```
[the remark this answers is not in the document]
Comment (ellisonmarks) [reply lib/ssh/setup-ssh-agent]: shebang and set added, and shellcheck resolved.

File: lib/ssh/setup-ssh-agent [hunk 5]
```diff
@@ -0,0 +1,9 @@
+sshdir=$(getent passwd $USER | cut -d: -f6)/.ssh
```
[the remark this answers is not in the document]
Comment (ellisonmarks) [reply lib/ssh/setup-ssh-agent]: addressed with the --set-home flag to sudo in default shell.

File: lib/ssh/setup-ssh-agent (hunk 2 above)
Comment (ellisonmarks) [reply lib/ssh/setup-ssh-agent]: switched to reading the key from stdin.

File: lib/ssh/setup-ssh-agent (hunk 1 above)
Comment (ellisonmarks) [reply lib/ssh/setup-ssh-agent]: removed all this in favor of reading the key from stdin.

File: lib/ssh/setup-ssh-agent (hunk 2 above)
Comment (ellisonmarks) [reply lib/ssh/setup-ssh-agent]: adopted this organization.

**Pull request merged**